llvm/lib/Support/Compression.cpp

```diff
@@ -202,6 +207,71 @@ Error zstd::decompress(ArrayRef<uint8_t> Input,
   return E;
 }
 
+static void zstdContextFree(ZSTD_DCtx *Ctx) {
+  size_t RC = ::ZSTD_freeDCtx(Ctx);
+  assert(::ZSTD_isError(RC) == 0);
+}
+
+Error zstd::decompress(ArrayRef<uint8_t> Input,
+                       SmallVectorImpl<uint8_t> &Output) {
+  size_t Res = 0;
+  std::unique_ptr<::ZSTD_DCtx, void (*)(::ZSTD_DCtx *)> Ctx(::ZSTD_createDCtx(),
+                                                            zstdContextFree);
+  Res = ::ZSTD_initDStream(&*Ctx);
+  if (::ZSTD_isError(Res))
+    return make_error<StringError>(::ZSTD_getErrorName(Res),
+                                   inconvertibleErrorCode());
+
+  // Create a buffer, this will be filled up repeatedly by
+  // ZSTD_decompressStream and appended to Output
+  SmallVector<char> Buffer;
+  Buffer.resize_for_overwrite(4096);
+
+  ::ZSTD_inBuffer InputBuffer = {
+      .src = Input.data(), .size = Input.size(), .pos = 0};
+  ::ZSTD_outBuffer OutputBuffer = {
+      .dst = Buffer.data(), .size = Buffer.size(), .pos = 0};
+
+  // First loop, read the entirety of `InputBuffer`. Stop when
+  // ZSTD_decompressedStream reports pos to be equal to size. This signals that
+  // all the data has been read and is at least in the internal state of the
+  // decompressor.
+  while (InputBuffer.pos < InputBuffer.size) {
+    Res = ::ZSTD_decompressStream(&*Ctx, &OutputBuffer, &InputBuffer);
+    if (::ZSTD_isError(Res))
+      return make_error<StringError>(::ZSTD_getErrorName(Res),
+                                     inconvertibleErrorCode());
+
+    // Check if the bufffer was filled, this might not happen in all iterations
+    // because the data read might still reside in the internal state of the
+    // decompressor
+    if (OutputBuffer.pos > 0) {
+      Output.append(Buffer.begin(), Buffer.begin() + OutputBuffer.pos);
+      OutputBuffer.pos = 0;
+    }
+  }
+
+  // Second loop, in this one there isn't any data to read, but there still
+  // might be data to be outputted from the decompressor internal state
+  while (true) {
```

Comment:
The second loop role seems to flush out internal data. Can you actually have more than one iteration of the second loop? Seems very weird.

Reply:
The API is opaque on this regard, the only way to know is to keep going until it returns that it has nothing more to write

Reply:
I feel like we need an helper function returning a `bool`.

Reply:
I'm not sure I'm following.
What would this helper function look like?

Reply:
Something like:

```c++
Expected<bool> process(...) {
  size_t Res = ::ZSTD_decompressStream(Ctx, &ZSTDOutput, &ZSTDInput);
  if (::ZSTD_isError(Res)) {
    return make_error<StringError>(::ZSTD_getErrorName(Res),
                                  inconvertibleErrorCode());
  }

  if (ZSTDOutput.pos > 0) {
    Output.append(Buffer.begin(), Buffer.begin() + ZSTDOutput.pos);
    ZSTDOutput.pos = 0;
    return true;
  }
  return false;
}
```

The two loops should then become more intelligible.

Reply:
Implemented